tools/protocol-schema-check/src/main.rs

```diff
@@ -0,0 +1,107 @@
+use near_primitives::hash::CryptoHash;
+use near_structs_checker::ProtocolStruct;
+use near_structs_checker_core::ProtocolStructInfo;
+use std::collections::hash_map::DefaultHasher;
+use std::collections::{HashMap, HashSet};
+use std::fs;
+use std::hash::{Hash, Hasher};
+// #[derive(Copy, Clone, Debug, ProtocolStruct)]
+// pub struct CryptoHash(pub [u8; 33]);
+
+fn compute_hash(
+    info: &ProtocolStructInfo,
+    structs: &HashMap<&'static str, &'static ProtocolStructInfo>,
+) -> u64 {
+    let mut hasher = DefaultHasher::new();
+    match info {
+        ProtocolStructInfo::Struct { name, fields } => {
+            name.hash(&mut hasher);
+            for (field_name, field_type) in *fields {
+                field_name.hash(&mut hasher);
+                compute_type_hash(field_type, structs, &mut hasher);
+            }
+        }
+        ProtocolStructInfo::Enum { name, variants } => {
+            name.hash(&mut hasher);
+            for (variant_name, variant_fields) in *variants {
+                variant_name.hash(&mut hasher);
+                if let Some(fields) = variant_fields {
+                    for (field_name, field_type) in *fields {
+                        // println!("Field: {} {}", field_name, field_type);
+                        field_name.hash(&mut hasher);
+                        compute_type_hash(field_type, structs, &mut hasher);
+                    }
+                }
+            }
+        }
+    }
+    hasher.finish()
+}
+
+fn compute_type_hash(
+    ty: &str,
+    structs: &HashMap<&'static str, &'static ProtocolStructInfo>,
+    hasher: &mut DefaultHasher,
+) {
+    // println!("Computing hash for {}", ty);
+    if let Some(nested_info) = structs.get(ty) {
+        compute_hash(nested_info, structs).hash(hasher);
+    } else {
+        // Handle generic types
+        let parts: Vec<&str> = ty.split('<').collect();
```

Comment:
May make sense to store a full type AST from procmacros so that you don't need to parse strings out in here. e.g. I think this code is currently invalid when there are nested generic types.

Reply:
An alternative would be to never allow generic schemas (that way you also don't need to worry about "generic" fields, as they would be required to use concrete types.)

Reply:
made a TODO